File: app/src/main/java/fr/android/androidexercises/RecyclerAdapter.java
package fr.android.androidexercises;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

/**
 * Created by anael on 09/11/2017.
 */

public class RecyclerAdapter<T> extends RecyclerView.Adapter {
    private final List<T> collection;
    private final LayoutInflater inflater;
    private int resource;

    public RecyclerAdapter(LayoutInflater inflater, List<T> collection, int resource) {
        this.collection = collection;
        this.inflater = inflater;
        this.resource = resource;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(inflater.inflate(this.resource, parent, false));
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        ((ItemView)holder.itemView).bind(this.collection.get(position));
    }

    @Override
    public int getItemCount() {
        return collection.size();
    }

    private class ViewHolder extends RecyclerView.ViewHolder {
        public ViewHolder(View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/fr/android/androidexercises/RecyclerAdapter.kt
package fr.android.androidexercises

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup


class RecyclerAdapter(private val inflater: LayoutInflater, private val collection: List<Book>, private val resource: Int) : RecyclerView.Adapter<RecyclerAdapter.ViewHolder>() {

    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
        (holder!!.itemView as BookItemView).bind(this.collection[position])    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(inflater.inflate(this.resource, parent, false))
    }

    override fun getItemCount(): Int {
        return collection.size
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}